### PR TITLE
Add golangci lint

### PR DESCRIPTION
**Pull request by usernamedt**
I propose to add golangci-lint with following checks enabled:
    - bodyclose
    - dupl
    - errcheck
    - funlen
    - gocritic
    - gocyclo
    - gofmt
    - goimports
    - govet
    - ineffassign
    - interfacer
    - misspell
    - nakedret
    - scopelint
    - staticcheck
    - structcheck
    - typecheck
    - unconvert
    - unparam
    - unused
    - varcheck
    
These mostly style-related checks are temporarily disabled, but I think it would be nice to enable them in the future:
    - golint
    - lll
    - godox
    - deadcode
    - stylecheck
    - gosimple
    - whitespace

### CODE REVIEW

**Comment by usernamedt** [comment]
Still need to fix these errors to make CI green:
```
internal/crypto/awskms/crypter.go:52:38: Error return value of `crypter.SymmetricKey.SetEncryptedKey` is not checked (errcheck)
        crypter.SymmetricKey.SetEncryptedKey(encryptedSymmetricKey)
                                            ^
internal/walparser/wal_parser.go:112:1: cyclomatic complexity 17 of func `(*WalParser).parsePage` is high (> 15) (gocyclo)
func (parser *WalParser) parsePage(reader io.Reader) (*XLogPage, error) {
^
internal/bguploader.go:102:26: Error return value of `b.workerCountSem.Acquire` is not checked (errcheck)
        b.workerCountSem.Acquire(context.TODO(), int64(b.maxParallelWorkers))
                                ^
internal/pagefile.go:207:17: Error return value of `file.Sync` is not checked (errcheck)
        defer file.Sync()
                       ^
internal/prefetch.go:207:13: Error return value of `os.MkdirAll` is not checked (errcheck)
        os.MkdirAll(runningLocation, 0755)
                   ^
internal/prefetch.go:215:12: Error return value of `os.Rename` is not checked (errcheck)
                os.Rename(oldPath, newPath)
                         ^
internal/queryRunner.go:195:19: Error return value of `tx.Rollback` is not checked (errcheck)
        defer tx.Rollback()
                         ^
internal/tar_interpreter_new.go:27:22: Error return value of `localFile.Sync` is not checked (errcheck)
        defer localFile.Sync()
                            ^
internal/backup_push_handler.go:71: Function 'createAndPushBackup' has too many statements (79 > 50) (funlen)
func createAndPushBackup(
internal/databases/sqlserver/blob/blob.go:134:1: cyclomatic complexity 17 of func `(*Index).PutBlockList` is high (> 15) (gocyclo)
func (idx *Index) PutBlockList(xblocklist *XBlockListIn) ([]string, error) {
^
internal/backup_fetch_handler.go:163:3: ineffectual assignment to `spec` (ineffassign)
                spec = sentinelDto.TablespaceSpec
                ^
```

File: internal/crypto/awskms/key.go [hunk 1]
```diff
@@ -45,7 +45,12 @@ func (symmetricKey *SymmetricKey) Encrypt() error {
 		kmsConfig = kmsConfig.WithRegion(symmetricKey.Region)
 	}
 
-	svc := kms.New(session.New(), kmsConfig)
+	kmsSession, err := session.NewSession()
```
Comment (usernamedt) [inline internal/crypto/awskms/key.go:48]: linter complained about deprecated New() function

File: internal/crypto/libsodium/reader.go [hunk 2]
```diff
@@ -100,8 +100,8 @@ func (reader *Reader) readNextChunk() (err error) {
 	returnCode := C.crypto_secretstream_xchacha20poly1305_pull(
 		&reader.state,
 		(*C.uchar)(&reader.out[0]),
-		(*C.ulonglong)(&outLen),
-		(*C.uchar)(&tag),
+		&outLen,
```
Comment (usernamedt) [inline internal/crypto/libsodium/reader.go:103]: Reason: linter complained about unnecesary cast

File: internal/crypto/openpgp/crypter.go [hunk 3]
```diff
@@ -64,7 +64,8 @@ func (crypter *Crypter) setupPubKey() error {
 		return nil
 	}
 
-	if crypter.IsUseArmoredKey {
+	switch {
```
Comment (usernamedt) [inline internal/crypto/openpgp/crypter.go:67]: Reason: linter complained about converting if-else chain to switch case

File: internal/databases/sqlserver/backup_push_handler.go [hunk 4]
```diff
@@ -71,19 +75,24 @@ func HandleBackupPush(dbnames []string, updateLatest bool, compression bool) {
 	tracelog.InfoLogger.Printf("backup finished")
 }
 
-func backupSingleDatabase(ctx context.Context, db *sql.DB, baseUrl string, dbname string, compression bool) error {
+func backupSingleItem(itemName string, ctx context.Context, db *sql.DB,
```
Comment (usernamedt) [inline internal/databases/sqlserver/backup_push_handler.go:78]: Reason: linter complained about duplicate functions: backupSingleDatabase / backupSingleLog

File: internal/databases/sqlserver/blob/util.go [hunk 5]
```diff
@@ -17,9 +16,8 @@ var ErrNotFound = errors.New("object not found")
 var ErrBadRequest = errors.New("invalid request")
 
 type Lease struct {
-	folder storage.Folder
```
Comment (usernamedt) [inline internal/databases/sqlserver/blob/util.go:20]: Reason: linter complained about unused fields in struct

File: internal/databases/sqlserver/sqlserver.go [hunk 6]
```diff
@@ -229,14 +229,3 @@ func uniq(src []string) []string {
 	}
 	return res
 }
-
-func convertPITRTime(untilDt string) (string, error) {
```
Comment (usernamedt) [inline internal/databases/sqlserver/sqlserver.go:233]: Reason: linter complained about unused func

File: internal/databases/mysql/backup_mark_handler.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ func MarkBackup(uploader *internal.Uploader, backupName string, toPermanent bool
 		meta.IsPermanent = toPermanent
 	}
 
-	metadataUploadObject, err := internal.GetMetadataUploadObject(backup.Name, meta)
+	metadataUploadObject, err := internal.GetMetadataUploadObject(backup.Name, &meta)
```
Comment (usernamedt) [inline internal/databases/mysql/backup_mark_handler.go:33]: Reason: linter complained about heavy metadata struct passed by value

File: internal/databases/mysql/binlog_fetch_handler.go [hunk 8]
```diff
@@ -43,16 +42,10 @@ func (ih *indexHandler) createIndexFile() error {
 }
 
 func HandleBinlogFetch(folder storage.Folder, backupName string, untilTs string) {
-	backup, err := internal.GetBackupByName(backupName, utility.BaseBackupPath, folder)
```
Comment (usernamedt) [inline internal/databases/mysql/binlog_fetch_handler.go:46]: Reason: linter complained about code duplication: HandleBinlogFetch / HandleBinlogReplay

File: cmd/mongo/oplog_replay.go [hunk 9]
```diff
@@ -51,10 +49,6 @@ type oplogReplayRunArgs struct {
 
 	oplogAlwaysUpsert    *bool
 	oplogApplicationMode *string
-
```
Comment (usernamedt) [inline cmd/mongo/oplog_replay.go:54]: Reason: linter complained about unused struct fields

File: internal/databases/mongo/archive/loader.go [hunk 10]
```diff
@@ -294,7 +294,8 @@ func NewStoragePurger(opts StorageSettings) (*StoragePurger, error) {
 // TODO: extract BackupLayout abstraction and provide DataPath(), SentinelPath(), Exists() methods
 func (sp *StoragePurger) DeleteBackups(backups []models.Backup) error {
 	keys := make([]string, 0, len(backups)*2)
-	for _, backup := range backups {
+	for idx := range backups {
+		backup := &backups[idx]
```
Comment (usernamedt) [inline internal/databases/mongo/archive/loader.go:298]: Reason: linter complained about large struct copying on each iteration

File: internal/asm/archive_status_manager_nop.go [hunk 11]
```diff
@@ -1,10 +1,6 @@
 package asm
 
-import "github.com/wal-g/wal-g/internal/fsutil"
-
-type NopASM struct {
```
Comment (usernamedt) [inline internal/asm/archive_status_manager_nop.go:5]: unused struct field

File: internal/backup_list_handler.go [hunk 12]
```diff
@@ -168,8 +168,9 @@ func writePrettyBackupListDetails(backupDetails []BackupDetail, output io.Writer
 	writer.SetOutputMirror(output)
 	defer writer.Render()
 	writer.AppendHeader(table.Row{"#", "Name", "Last modified", "WAL segment backup start", "Start time", "Finish time", "Hostname", "Datadir", "PG Version", "Start LSN", "Finish LSN", "Permanent"})
-	for i, b := range backupDetails {
-		writer.AppendRow(table.Row{i, b.BackupName, b.Time.Format(time.RFC850), b.WalFileName, b.StartTime.Format(time.RFC850), b.FinishTime.Format(time.RFC850), b.Hostname, b.DataDir, b.PgVersion, b.StartLsn, b.FinishLsn, b.IsPermanent})
+	for idx := range backupDetails {
+		b := &backupDetails[idx]
```
Comment (usernamedt) [inline internal/backup_list_handler.go:172]: Reason: linter complained about large struct copy on each iteration

File: internal/backup_mark.go [hunk 13]
```diff
@@ -76,7 +76,7 @@ func getMarkedPermanentBackupMetadata(baseBackupFolder storage.Folder, backupNam
 	// only return backups that we want to update
 	if !meta.IsPermanent {
 		meta.IsPermanent = true
-		metadataUploadObject, err := GetMetadataUploadObject(backup.Name, meta)
+		metadataUploadObject, err := GetMetadataUploadObject(backup.Name, &meta)
```
Comment (usernamedt) [inline internal/backup_mark.go:79]: Reason: linter complained about large struct passed by value

File: internal/bundle_files.go [hunk 14]
```diff
@@ -108,8 +108,9 @@ func newRelFileStatistics(conn *pgx.Conn) (RelFileStatistics, error) {
 	result := make(map[walparser.RelFileNode]PgRelationStat)
 	// CollectStatistics collects statistics for each relFileNode
 	for _, db := range databases {
+		dbName := db.name
```
Comment (usernamedt) [inline internal/bundle_files.go:111]: Reason: linter complained about potential bug

File: internal/delta_no.go [hunk 15]
```diff
@@ -19,12 +19,6 @@ func newDeltaNoFromFilename(filename string) (DeltaNo, error) {
 	return DeltaNo(no), err
 }
 
-func newDeltaNoFromFilenameNoError(filename string) DeltaNo {
```
Comment (usernamedt) [inline internal/delta_no.go:22]: Reason: linter complained about duplicate functions

File: internal/paged_file_verifier.go [hunk 16]
```diff
@@ -72,7 +72,7 @@ func calculateSum(checksum, value uint32) uint32 {
 // The checksum includes the block number (to detect the case where a page is
 // somehow moved to a different location), the page header (excluding the
 // checksum itself), and the page data.
-func pgChecksumPage(blockNo uint32, pageBytes [DatabasePageSize]byte) (uint16, error) {
```
Comment (usernamedt) [inline internal/paged_file_verifier.go:75]: Reason: linter complained about large array passed by value

File: internal/pagefile.go [hunk 17]
```diff
@@ -32,7 +32,6 @@ import (
 const (
 	DatabasePageSize            = int64(walparser.BlockSize)
 	sizeofInt32                 = 4
-	sizeofInt16                 = 2
```
Comment (usernamedt) [inline internal/pagefile.go:35]: Reason: linter complained about unused const (ignore?)

File: internal/wal_show_handler.go [hunk 18]
```diff
@@ -198,8 +198,7 @@ func getBackupsInRange(startSegmentName, endSegmentName string, timeline uint32,
 
 		// there we compare segments lexicographically, maybe it is wrong...
 		if timeline == backupTimeline && startSegment >= startSegmentName && endSegment <= endSegmentName {
-			filteredBackup := backup
-			filteredBackups = append(filteredBackups, &filteredBackup)
```
Comment (x4m) [inline internal/wal_show_handler.go:202]: It would be good to define var near usage.

**Review by x4m — APPROVED** [review]
PG and internal stuff look good to me.

File: internal/databases/sqlserver/backup_push_handler.go [hunk 19]
```diff
@@ -14,6 +15,9 @@ import (
 	"github.com/wal-g/wal-g/utility"
 )
 
+const DatabaseBackupItem = "DATABASE"
+const LogBackupItem = "LOG"
```
Comment (mialinx) [inline internal/databases/sqlserver/backup_push_handler.go:19]: Nope.  DATABASE and LOG backup are completely different operation.
SQL queires are just similar now by will diverge in the future.

File: internal/databases/sqlserver/backup_push_handler.go (hunk 4 above)
Comment (mialinx) [reply internal/databases/sqlserver/backup_push_handler.go:78]: Let's disable this linter.

File: internal/databases/sqlserver/blob/blob.go [hunk 20]
```diff
@@ -147,39 +147,9 @@ func (idx *Index) PutBlockList(xblocklist *XBlockListIn) ([]string, error) {
...
-					garbage = append(garbage, fmt.Sprintf("%s.%d", xb.ID, block.CommittedRev))
-				}
-				block.CommittedRev = block.UploadedRev
-				block.CommittedSize = block.UploadedSize
-				block.UploadedRev = 0
-				block.UploadedSize = 0
-			}
-		default:
-			panic(fmt.Sprintf("unexpected block mode: %s", xb.Mode))
+		err := checkBlockMode(block, xb, &garbage)
```
Comment (mialinx) [inline internal/databases/sqlserver/blob/blob.go:150]: nope. The readability suffers.

File: internal/databases/sqlserver/blob/server.go [hunk 21]
```diff
@@ -406,7 +406,9 @@ func (bs *Server) HandleBlockListGet(w http.ResponseWriter, req *http.Request) {
 	w.Header().Set("Content-Length", strconv.Itoa(len(data)))
 	w.Header().Set("Content-Type", "application/xml; charset=utf-8")
 	w.WriteHeader(http.StatusOK)
-	w.Write(data)
+	if _, err = w.Write(data); err != nil {
+		bs.returnError(w, req, err)
```
Comment (mialinx) [inline internal/databases/sqlserver/blob/server.go:410]: bs.returnError - sends HTTP headers.
We can't send headers after writing HTTP body

File: internal/databases/sqlserver/log_push_handler.go [hunk 22]
```diff
@@ -39,26 +36,9 @@ func HandleLogPush(dbnames []string, compression bool) {
 	logBackupName := generateLogBackupName()
 	baseUrl := getLogBackupUrl(logBackupName)
 	err = runParallel(func(dbname string) error {
-		return backupSingleLog(ctx, db, baseUrl, dbname, compression)
+		return backupSingleItem(LogBackupItem, ctx, db, baseUrl, dbname, compression)
```
Comment (mialinx) [inline internal/databases/sqlserver/log_push_handler.go:39]: nope

**Comment by mialinx** [comment]
MySQL looks good.
SQLServer need changes.

May be disable dupl linter ?

File: internal/databases/sqlserver/log_push_handler.go (hunk 22 above)
Comment (usernamedt) [reply internal/databases/sqlserver/log_push_handler.go]: fixed in b5ae4f4

File: internal/databases/sqlserver/backup_push_handler.go (hunk 19 above)
Comment (usernamedt) [reply internal/databases/sqlserver/backup_push_handler.go]: fixed in b5ae4f4

File: internal/databases/sqlserver/blob/server.go (hunk 21 above)
Comment (usernamedt) [reply internal/databases/sqlserver/blob/server.go]: fixed in 91de9e6

**Comment by usernamedt** [comment]
I propose to increase dupl linter threshold instead of disabling it so we can still detect some huge code duplication

File: internal/databases/sqlserver/blob/blob.go (hunk 20 above)
Comment (usernamedt) [reply internal/databases/sqlserver/blob/blob.go]: I removed the changes and added gocyclo ignore flag in 7437f84

**Comment by mialinx** [comment]
SQLServer ok

**Review by mialinx — APPROVED** [review]
MySQL & SQLServer are ok

**Review by perekalov — APPROVED** [review]

File: internal/databases/sqlserver/blob/blob.go [hunk 23]
```diff
@@ -336,9 +336,3 @@ func (idx *Index) GetCachedReader(folder storage.Folder, s Section) (io.ReadClos
 	}
 	return ioutil.NopCloser(bytes.NewReader(buf)), nil
 }
-
-func (idx *Index) debugBlocks() {
```
Comment (mialinx) [inline internal/databases/sqlserver/blob/blob.go:339]: эта функция нужна. используется при отладке.

**Review by mialinx — APPROVED** [review]

**Pull request merged**